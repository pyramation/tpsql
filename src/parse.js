import fs from 'fs';
import path from 'path';

export const parse = ({ lines, fnName }) => {
  const replaces = [];
  const commentExp = /^\s*--/;
  const replaceRegExp = /^\s*--\sreplace:\s(.*)\s(.*)\s(.*)$/;
  const requireRegExp = /^\s*--\srequires:\s(.*)$/;
  const deployRegExp = /^\s*--\sDeploy\s(.*)$/;
  const revertRegExp = /^\s*--\sRevert\s(.*)$/;
  const verifyRegExp = /^\s*--\sVerify\s(.*)$/;
  const hintRegexp = /^\s*--\shint:\s(.*)$/;
  const revertExp = /^\s*--\srevert:/;
  const verifyExp = /^\s*--\sverify:/;

  const deps = [];
  let deploys = [];
  let type;
  let mode = 'deploy';
  const reverts = [];
  const verifys = [];
  lines = lines
    .map((line, l) => {
      if (replaceRegExp.test(lines[l])) {
        const [m, find, replace, type] = line.match(replaceRegExp);
        replaces.push({ find, replace, type });
        return '';
      } else if (hintRegexp.test(line)) {
        const [m, type] = line.match(hintRegexp);
        return '';
      } else if (requireRegExp.test(line)) {
        const [m, stuff] = line.match(requireRegExp);
        deps.push(stuff.trim().split(' '));
        return '';
      } else if (deployRegExp.test(line)) {
        const [m, stuff] = line.match(deployRegExp);
        deploys = stuff.trim();
        type = 'Deploy';
        return '';
      } else if (revertRegExp.test(line)) {
        const [m, stuff] = line.match(revertRegExp);
        deploys = stuff.trim();
        type = 'Revert';
        return '';
      } else if (verifyRegExp.test(line)) {
        const [m, stuff] = line.match(verifyRegExp);
        type = 'Verify';
        deploys = stuff.trim();
        return '';
      } else if (revertExp.test(line)) {
        mode = 'revert';
        return '';
      } else if (verifyExp.test(line)) {
        mode = 'verify';
        return '';
        // commentExp MUST go LAST for new comment helpers
      } else if (commentExp.test(line)) {
        return '';
      }

      if (mode == 'revert') {
        reverts.push(line);
      } else if (mode === 'verify') {
        verifys.push(line);
      } else return line;
      return '';
    })
    .filter((el) => el !== '');

  if (!lines.length) {
    return;
  }

  if (lines[0].trim() === 'BEGIN;') {
    lines.shift();
  }
  if (lines[lines.length - 1].trim() === 'COMMIT;') {
    lines.pop();
  }

  const args = replaces.map(({ replace }) => replace);
  const orderedReplaces = replaces
    .map((a) => a)
    .sort(function (a, b) {
      // ASC  -> a.length - b.length
      // DESC -> b.length - a.length
      return b.find.length - a.find.length;
    });

  const orderOfArgs = replaces.reduce((m, v, i) => {
    m[v.find] = i;
    return m;
  }, {});

  deps.forEach((dep) => {
    const [m, ...rest] = dep;
    rest.forEach((el) => {
      if (!args.includes(el)) {
        if (el.match(/^'[^']+'$/)) {
          // console.log('Found a STRING '+el);
        } else {
          console.error({ el });
          console.error({ args });
          console.error(`bad dep ${el} in ${fnName}`);
          process.exit(1);
        }
      }
    });
  });

  return {
    lines,
    deploys,
    verifys,
    reverts,
    orderedReplaces,
    replaces,
    deps,
    type
  };
};

export const parseFile = (sqlfile) => {
  const fnName = path.basename(sqlfile).replace(path.extname(sqlfile), '');
  const lines = fs.readFileSync(sqlfile).toString().split('\n');
  return parse({ lines, fnName });
};

export const parseFiles = (files) => {
  return files.reduce((m, file) => {
    m[file] = parseFile(file);
    return m;
  }, {});
};
