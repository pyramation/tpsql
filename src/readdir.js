import { sync as glob } from 'glob';

export const readdir = (dirname = process.cwd()) => {
  return glob(dirname + '/**/*.sql');
};
