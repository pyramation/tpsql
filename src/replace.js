import format from 'pg-format';

const createFormatString = ({ lines, replaces, orderedReplaces }) => {
  const orderOfArgs = replaces.reduce((m, v, i) => {
    m[v.find] = i;
    return m;
  }, {});

  const newlines = [];
  for (let l = 0; l < lines.length; l++) {
    let line = lines[l].replace(/%/g, '%%');
    let line2 = lines[l].replace(/%/g, '%%');
    for (let r = 0; r < orderedReplaces.length; r++) {
      const { find, replace, type } = orderedReplaces[r];
      const properIndex = orderOfArgs[find];
      line2 = line2.replace(new RegExp(find, 'g'), replace);
      line = line.replace(new RegExp(find, 'g'), `%${properIndex + 1}$${type}`);
    }
    newlines.push(line);
  }
  return newlines.join('\n');
};

const getDepInfo = ({ deps, orderedReplaces, variables }) => {
  const dps = [];
  for (const d of deps.flat()) {
    const dep = replacer({ string: d, orderedReplaces, variables });
    dps.push(dep);
  }
  return dps;
};

const replacer = ({ string, orderedReplaces, variables }) => {
  let str = string;
  for (const r of orderedReplaces) {
    const { find, replace } = r;
    str = str.replace(find, replace);
  }
  for (const r of orderedReplaces) {
    const { replace } = r;
    str = str.replace(replace, variables[replace]);
  }
  return str;
};

export const getFormat = (obj) => {
  return Object.keys(obj).reduce((m, k) => {
    const { lines, replaces, orderedReplaces } = obj[k];
    obj[k].format = createFormatString({ lines, replaces, orderedReplaces });
    m[k] = obj[k];
    return m;
  }, {});
};

export const getDeps = (obj, variables) => {
  return Object.keys(obj).reduce((m, k) => {
    const { deploys, deps, replaces, orderedReplaces } = obj[k];

    obj[k].info = {
      deploy: replacer({
        string: deploys,
        orderedReplaces,
        variables
      }),
      deps: getDepInfo({ deps, replaces, orderedReplaces, variables })
    };

    m[k] = obj[k];
    return m;
  }, {});
};

export const getArgs = (obj) => {
  return Object.keys(obj).reduce((m, k) => {
    const { deps, replaces, orderedReplaces } = obj[k];
    obj[k].args = replaces.map(({ replace }) => replace);
    m[k] = obj[k];
    return m;
  }, {});
};

const getOrderedArgs = ({ orderedReplaces, variables }) => {
  return orderedReplaces.map(({ replace }) => {
    if (!variables) {
      throw new Error('missing variables');
    }
    if (!variables.hasOwnProperty(replace)) {
      throw new Error('missing ' + replace);
    }
    return variables[replace];
  });
};

export const getReplaced = (obj, variables) => {
  return Object.keys(obj).reduce((m, k) => {
    const { orderedReplaces } = obj[k];

    obj[k].replaced = format(
      obj[k].format,
      ...getOrderedArgs({ orderedReplaces, variables })
    );

    m[k] = obj[k];
    return m;
  }, {});
};

export const replace = (obj, variables) => {
  obj = getFormat(obj);
  obj = getDeps(obj, variables);
  obj = getArgs(obj);
  obj = getReplaced(obj, variables);
  return obj;
};
