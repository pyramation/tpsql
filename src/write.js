import { dirname, basename } from 'path';
import { join } from 'path';
import { writeFileSync } from 'fs';
import { sync as mkdirp } from 'mkdirp';

export const write = (obj, outdir) => {
  return Object.keys(obj).reduce((m, k) => {
    const {
      info: { deploy, deps },
      replaced
    } = obj[k];
    const mod = deploy.split(' ')[0];
    const modulename = basename(mod) + '.sql';
    const dir = dirname(mod);

    const files = {
      deploy: join(outdir, 'deploy', dir, modulename),
      revert: join(outdir, 'revert', dir, modulename),
      verify: join(outdir, 'verify', dir, modulename)
    };

    let str = `-- Deploy ${mod} to pg
  
${deps.map((dep) => `-- requires: ${dep}`).join('\n')}
  
BEGIN;
  
${replaced}
  
END;
      `;

    mkdirp(dirname(files.deploy));
    writeFileSync(files.deploy, str);

    //

    str = `-- Revert ${mod} from pg
  
BEGIN;
  
END;
      `;

    mkdirp(dirname(files.revert));
    writeFileSync(files.revert, str);

    //

    str = `-- Verify ${mod} on pg
  
BEGIN;
  
END;
      `;

    mkdirp(dirname(files.verify));
    writeFileSync(files.verify, str);

    m[k] = obj[k];
    return m;
  }, {});
};
