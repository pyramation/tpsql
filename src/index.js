export * from './readdir';
export * from './parse';
export * from './replace';
export * from './write';

import { getAllArgsFn, replace } from './replace';
import { readdir } from './readdir';
import { parseFiles } from './parse';

import { write as _write } from './write';

export const templatize = (deployDir, outputDir, variables) => {
  const files = readdir(deployDir);
  const obj = parseFiles(files);
  const result = replace(obj, variables);
  _write(result, outputDir);
};

export class Templatizer {
  constructor(deployDir) {
    this.deployDir = deployDir;
    this.files = readdir(deployDir);
    this.obj = parseFiles(this.files);
  }
  variables() {
    return getAllArgsFn(this.obj);
  }
  write(outputDir, variables) {
    const result = replace(this.obj, variables);
    _write(result, outputDir);
    return this;
  }
}
