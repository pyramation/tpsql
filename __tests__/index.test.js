import { readdir, parseFiles, getAllArgs } from '../src';
import path from 'path';
import { tmpdir } from 'os';
import { randomBytes } from 'crypto';
import mkdirp from 'mkdirp';
import { replace, write, templatize, getAllArgsFn, Templatizer } from '../src';

const dir = randomBytes(20).toString('hex');

const packagesDir = path.join(tmpdir(), dir);
mkdirp.sync(packagesDir);

// NOTE: only support deploy currently... we do have the types.. but who cares about verify/revert rn.

const fixtures = path.join(__dirname, '/../__fixtures__/achievements/deploy');
const rinse = (path) => path.split('__fixtures__')[1];

it('replace', () => {
  templatize(fixtures, packagesDir, {
    private_schema: 'my-private_schema',
    public_schema: 'the_public_schema'
  });
  const out = readdir(packagesDir);
  // console.log(out);
});

it('replace', () => {
  const files = readdir(fixtures);
  let obj = parseFiles(files);
  obj = getAllArgsFn(obj);

  console.log(obj);
});

it('Templatizer', () => {
  const client = new Templatizer(fixtures);
  expect(client.variables()).toMatchSnapshot();
  client.write(packagesDir, {
    private_schema: 'my-private_schema',
    public_schema: 'the_public_schema'
  });
});
