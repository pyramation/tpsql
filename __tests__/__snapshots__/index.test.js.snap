// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Templatizer 1`] = `
Array [
  "private_schema",
  "public_schema",
]
`;
